src/arch/x86/x86_64/src/bzimage.rs

```diff
@@ -0,0 +1,293 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#![allow(non_camel_case_types)]
+use core::fmt::Write;
+use core::intrinsics::{copy, transmute};
+use core::mem::size_of;
+use model::{Driver, Result};
+use wrappers::{Memory, SectionReader};
+pub type EntryPoint = unsafe extern "C" fn(r0: usize, dtb: usize);
+//use std::mem::{size_of, transmute, zeroed};
+
+pub const E820_RAM: u32 = 1;
+pub const E820_RESERVED: u32 = 2;
+pub const E820_ACPI: u32 = 3;
+pub const E820_NVS: u32 = 4;
+pub const E820_UNUSABLE: u32 = 5;
+
+// TODO define these somewhere.
+const PAGE_SIZE: usize = 4096;
+
+#[repr(C, packed)]
+#[derive(Default, Copy, Clone)]
+pub struct E820Entry {
+    addr: u64,
+    size: u64,
+    r#type: u32,
+}
+
+pub const E820_MAX: usize = 128;
+#[repr(C, packed)]
+pub struct BootParams {
+    screen_info: [u8; 0x040],             // 0x000
+    apm_bios_info: [u8; 0x054 - 0x040],   // 0x040
+    _pad2: [u8; 4],                       // 0x054
+    tboot_addr: u64,                      // 0x058
+    ist_info: [u8; 0x070 - 0x060],        // 0x060
+    acpi_rsdp_addr: [u8; 0x078 - 0x070],  // 0x070
+    _pad3: [u8; 8],                       // 0x078
+    hd0_info: [u8; 0x090 - 0x080],        // 0x080 /* obsolete! */
```

Comment:
I think the idiomatic Rust way is to use `#[deprecated]` instead of `obsolete!`.

Reply:
The obsolete is from the original, so should probably not change.